core/src/main/java/org/apache/calcite/avatica/util/SecurityUtils.java

```diff
@@ -0,0 +1,266 @@
+   * </p>
+   * @param permission the permission to check
+   * @throws SecurityException if the permission check fails
+   */
+  public static void checkPermission(Permission permission) throws SecurityException {
+    Object securityManager = SecurityUtils.getSecurityManager();
+    if (securityManager == null) {
+      return;
+    }
+    try {
+      securityManager.getClass().getMethod("checkPermission").invoke(securityManager,
+          permission);
+    } catch (SecurityException x) {
+      throw x;
+    } catch (Throwable ignored) {
+    }
+  }
+
+  /**
+   * <p>
+   * Runs the given action with the calling context restricted to just the calling frame, not all
+   * the frames in the stack.
+   * </p>
+   * @param action the action to run
+   * @return the result of running the action
+   * @param <T> the type of the result
+   */
+  public static <T> T doPrivileged(PrivilegedAction<T> action) {
+    // Keep this method short and inlineable.
+    if (DO_PRIVILEGED == null) {
+      return action.run();
+    }
+    return doPrivileged(DO_PRIVILEGED, action);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <T> T doPrivileged(MethodHandle doPrivileged, PrivilegedAction<T> action) {
+    try {
+      return (T) doPrivileged.invoke(action);
+    } catch (RuntimeException | Error x) {
+      throw x;
+    } catch (Throwable x) {
+      // Unlikely to happen, as we catch Errors and RuntimeExceptions above, and doPrivileged
+      // does not throw checked exceptions
+      throw new RuntimeException(x);
+    }
+  }
+
+  /**
+   * <p>
+   * Runs the given action as the given subject.
+   * </p>
+   * @param subject the subject this action runs as
+   * @param action the action to run
+   * @return the result of the action
+   * @param <T> the type of the result
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T callAs(Subject subject, Callable<T> action) {
+    try {
+      if (CALL_AS == null) {
+        throw new RuntimeException(
+            "Was unable to run either of Subject.callAs() or Subject.doAs()");
+      }
+      return (T) CALL_AS.invoke(subject, action);
+    } catch (RuntimeException | Error x) {
+      throw x;
+    } catch (Throwable x) {
+      throw new CompletionException(x);
+    }
+  }
+
+  /**
+   * <p>
+   * Gets the current subject
+   * </p>
+   * @return the current subject
+   */
+  @SuppressWarnings("unchecked")
+  public static Subject currentSubject() {
+    if (CURRENT == null) {
+      throw new RuntimeException(
+          "Was unable to run either of Subject.current() or Subject.getSubject()");
+    }
+    try {
+      MethodHandle methodHandle = CURRENT;
+      return (Subject) methodHandle.invoke();
+    } catch (Throwable x) {
+      throw new RuntimeException("Error while trying to get the current user", x);
+    }
+
+  }
+
+  private static <T> PrivilegedAction<T> callableToPrivilegedAction(Callable<T> callable) {
```

Comment:
Not investigated the wider context here... but I do wonder if this should map to `PrivilegedExceptionAction`?

Reply:
The current code does not PrivilegedExceptionAction, so we don't need to implement that functionality.

Reply:
What I mean here is Callable<?> throws checked exceptions but PrivilegedAction doesn't. In that sense:

_`PrivilegedAction<T>` is to `Supplier<T>` as `PrivilegedExceptionAction<T>` is to `Callable<T>`_